### ISSUE TITLE
Deleted branches still show up in the branches dropdown

### ISSUE THREAD

**Issue by IdanAdar** [desktop/desktop#5893]
## Description
I've created a branch via GitHub Desktop. Now I'm done with it and so I've deleted it from the GitHub.com UI.

In GitHub Desktop though it's "still there". 
How do I sync the list of branches? Does it sync at all?

## Version

* GitHub Desktop: 1.4.3 beta 1
* Operating system: macOS 10.14

## Steps to Reproduce
1. Create a branch via GitHub Desktop
2. Delete it from the GitHub.com UI

### Expected Behavior
For the branch to not be available in GitHub Desktop

### Actual Behavior
The branch is available in GitHub Desktop

**Comment by billygriffin** [comment]
Thanks @IdanAdar! We totally agree that it would be really helpful for Desktop to prune branches in Desktop that have been deleted on the remote. We're tracking this in #750 and we'd love for someone from the community to make a PR for this if possible.

In the meantime, you can delete branches individually from the menu bar, Branch -> Delete. Thanks again for the request.

**Comment by jamesjwilson** [comment]
Has this not been resolved yet?  I just pruned a bunch of branches and they still show up in GitHub Desltop.

**Comment by billygriffin** [comment]
@jamesjwilson It's only on the beta version for the time being. Since it's a potentially destructive action, we want to ensure it has plenty of time on beta to make sure it's solid and reliable and we're not accidentally deleting anyone's branch that has work on it that's not merged in, for example.

Additionally, when it does ship to production, it'll wait 14 days before being pruned in Desktop because we've found that people reuse the same branch names in many cases and don't want to have to recreate them each time if they're frequently used.

**Comment by jamesjwilson** [comment]
@billygriffin Ahh I misunderstood the request.  I delete everything manually via Git Bash and they still show up as branches for me in GitHub Desktop, I figured it should read from my repository instead of attempting to delete it by itself.  Hopefully, that will be a feature we can turn off, that does sound a bit scary.  I was just surprised it was seeing branches I had deleted already.

**Comment by billygriffin** [comment]
@jamesjwilson That's strange, would you mind opening a new issue if you're reliably seeing that behavior reproduced?

Here's my progression.

See branch (`delete-this-branch`):

<img width="410" alt="screen shot 2019-02-21 at 11 52 41 am" src="https://user-images.githubusercontent.com/5091167/53193904-675a8a00-35cf-11e9-95d0-fea688eacb90.png">

Delete the branch in bash:

<img width="496" alt="screen shot 2019-02-21 at 11 54 43 am" src="https://user-images.githubusercontent.com/5091167/53193947-7f320e00-35cf-11e9-832e-5c65300cff0a.png">

Go back to Desktop, don't see branch (`delete-this-branch`):

<img width="410" alt="screen shot 2019-02-21 at 11 53 19 am" src="https://user-images.githubusercontent.com/5091167/53193987-983abf00-35cf-11e9-8b38-044239a90bfe.png">

**Comment by jayantnd** [comment]
faced the same issue today.
created a branch in Github Desktop. created a PR, merged and then deleted the branch

But still i see the branch in desktop app after sync.

Mac OS 10.15.3
Github Desktop 2.4.0

**Comment by billygriffin** [comment]
@jayantnd The pruning only happens for branches that are older than 14 days and have no new commits since the merge. This was done so if people reuse the same branches, they still have access to them without having to recreate them each time.

**Comment by jbevarts** [comment]
I like deleting and recreating because it allows me to branch off of the most updated version of parent branch.  It might not be the most efficient flow, but i've run into this 14 day window plenty.

**Comment by bschaatsbergen** [comment]
fix ples

**Comment by JosephTLyons** [comment]
Really would love this feature in Desktop.  With our branching model, I have so many branches that pile up that it becomes overwhelming to manually delete them, even if on a fairly regular basis.  I would love some automated process for this.

**Comment by fluffy-cakes** [comment]
....and this is not fixed, because?

**Comment by sidfadnis** [comment]
glad there is a thread for this. i am newbie to github, and was going crazy if I was doing something wrong.

**Comment by ShaneKeney** [comment]
Is there an estimated time frame for this?

**Comment by I-Valchev** [comment]
please fix this...

**Comment by fluffy-cakes** [comment]
Just an FYI; I ran this on my machine and it worked a treat; https://medium.com/@kcmueller/delete-local-git-branches-that-were-deleted-on-remote-repository-b596b71b530c

However, I used the `-D` instead of lower case; this way it actually deletes the local old branches.

I have a Windows machine and I have WSL installed, so I just ran it in the same directory as my repo and it worked a treat.

**Comment by I-Valchev** [comment]
@fluffy-cakes thanks, but doesn't seem to work on mac:

```
git branch -vv | grep ': gone]'|  grep -v "\*" | awk '{ print $1; }' | xargs -r git branch -d
xargs: illegal option -- r
usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
             [-L number] [-n number [-x]] [-P maxprocs] [-s size]
             [utility [argument ...]]
```

**Comment by elliot-jh** [comment]
@I-Valchev macos terminal doesn't need the -r flag: 
`git branch -vv | grep ': gone]'| grep -v "\*"| awk '{print $1}'| xargs git branch -D `
should work (note the -D flag: this will force the branches to be deleted). This works for me on zsh.

**Comment by osmac984** [comment]
> @jamesjwilson That's strange, would you mind opening a new issue if you're reliably seeing that behavior reproduced?
> 
> Here's my progression.
> 
> See branch (`delete-this-branch`):
> 
> <img alt="screen shot 2019-02-21 at 11 52 41 am" width="410" src="https://user-images.githubusercontent.com/5091167/53193904-675a8a00-35cf-11e9-95d0-fea688eacb90.png">
> 
> Delete the branch in bash:
> 
> <img alt="screen shot 2019-02-21 at 11 54 43 am" width="496" src="https://user-images.githubusercontent.com/5091167/53193947-7f320e00-35cf-11e9-832e-5c65300cff0a.png">
> 
> Go back to Desktop, don't see branch (`delete-this-branch`):
> 
> <img alt="screen shot 2019-02-21 at 11 53 19 am" width="410" src="https://user-images.githubusercontent.com/5091167/53193987-983abf00-35cf-11e9-8b38-044239a90bfe.png">

I was having same issue, but this example working with CLI helped me to clean up my branch list

**Comment by I-Valchev** [comment]
@elliot-jh thanks, that does work indeed :)

**Comment by blockedhead** [comment]
Another way to handle his problem could be the followig:

1. Make sure that you've pushed everything to origin so you can retrieve it.

2.(GitHub Desktop) Remove the repository
![image](https://user-images.githubusercontent.com/73295213/103651174-751a1e00-4f61-11eb-92f8-59343e6b39a9.png)

3.(GitHub Online) Go into your repository and open the Code in GitHub Desktop 
![image](https://user-images.githubusercontent.com/73295213/103651285-a561bc80-4f61-11eb-9d7d-eac4e5a682b1.png)

I hope I've helped some people. Let me please know if there is any problem with this way. I am fairly new to GitHub.

**Comment by darcyrailip** [comment]
Kind of surprised this hasn't been resolved in over two years. Merging/deleting branches from PRs is all-too-common, and I have all these "ghost" branches in my GitHub Desktop.

Even a manual "Refresh" button like exists on the Pull Requests tab would be great.

**Comment by KingOtto123** [comment]
Adding to this, I see the same issue after merging a branch by pull request

Before, had a branch "A49". Merged that branch into my "maindev" branch via pull request on website. Online, the branch has disappeared (as it should), as the pull request was completed.

However, the "A49" branch still shows on GitHub desktop. On the right-hand side (main window), it even suggests to "Publish your branch", saying that it cannot be found on the remote.

Looks like the online pull request has resulted in the offline branch being virtually a separate branch now, that I have to manually delete offline. What am I doing wrong, here?

**Comment by b21985164** [comment]
Another year, is there a simpler solution to remove all ghost branches from github desktop yet?

**Comment by dev-yeet** [comment]
I'll chime in because this is very annoying, even with version 3.0 this is a problem

**Comment by ossyraabv** [comment]
Yep why is this such a longstanding issue?

**Comment by AheadWind** [comment]
Still an issue

**Comment by hidetsuna** [comment]
Would be nice to have that fixed

**Comment by Pchatain** [comment]
Still an issue as far as I can tell

**Comment by misterrodg** [comment]
Definitely still an issue, though relatively minor. Just verified on a branch I had merged on the website. Was still there after I had fetched origin.

**Comment by AlexanderMishutkinP** [comment]
October. Still an issue
Especially, when a large team of developers work together and produce 10-20 branches each week

**Comment by NORR-KanenV** [comment]
Keeping this post alive - No fix yet:(

**Comment by Chalderrr** [comment]
Hi there, this is still an issue for me in November.

**Comment by FelixConni** [comment]
Damm how is this still not fixed after 4years?  
I know i can do it through command but I use the gui so I can save some time typing it out.   
why don't you just delete the local branches if there is no commits to be done and pull from repo so it always grabs the latest branches?

**Comment by mlewicki103** [comment]
This is still an issue and I'd love if it got fixed.

**Comment by kiketordera** [comment]
Yes, is super annoying and shouldn't be difficult to fix

**Comment by wyteco** [comment]
Totally agree on this feature. I think we could generalise this however w/o needing to handle deletion periods (as proposed by @billygriffin 14 days) and in result behaviour differences from person to person and team to team:

- Add a label next to each branch in the branches list showing the status of the branch. Statuses could be "unpublished", "published" (no label), "pull request" (not sure on this one), "deleted".
This could be a great addition in itself as I—and maybe other people (haven't skimmed through other issues yet)—sometimes feel unsure about whether I already published another branch. Having to checkout that branch and then checking would be a simplified with a quick look at the overview.

- Add a multi-select feature on the branches list to perform certain actions in bulk. This could be the "delete" action but might also be extended in the future with "publish" for example.

I think that these two features make for a great addition to GitHub Desktop and would also address the pain-point discussed in this issue. A status indicator combined with bulk delete would suit this well IMO.

I would love to hear your opinions on it. We should also reference other issues addressed by these two features and/ or move this topic to its own discussion/ issue.

**Comment by wyteco** [comment]
Addition to my previous comment:

A challenge with the first feature is that we need to differentiate between a **not yet** used (unpublished) and a **not anymore** used branch (deleted). Also @billygriffin as you already said, we need to make sure the status is accurate—even across GitHub Desktop installations (so no extra local DB keeping track, we obviously need to get this done with git as the single source of truth)

**Comment by robrondon** [comment]
this is still an issue.. it should be reopened

**Comment by IbenBru** [comment]
4.5 years later...

**Comment by huddletron** [comment]
still an issue. should be reopened

**Comment by tylerdonovan123** [comment]
Still an issue, needs reopened

**Comment by jnhwkim** [comment]
Still an issue.

**Comment by mjds001** [comment]
would be great to have a fix for this

**Comment by Cybdyn-Systems** [comment]
Can someone fix this? It's causes a lot new people we hire who don't know how to use Git to get extremely confused and re-commit to already deleted branches.

**Comment by modernmusician** [comment]
i'm running into this issue now. can we please reopen and improve the user interface for this?

thank you 🙏

**Comment by Distortedlogic** [comment]
still an issue

**Comment by Distortedlogic** [comment]
allow users to delete without the 14 days if they want in a setting or something, sheesh

**Comment by JorisHageman** [comment]
Still in issue, this should be fixed. Maybe an option somewhere "Sync branches with remote"

**Comment by StephanBis** [comment]
How is this, after almost 5 years, still not fixed. This should be re-opened and fixed.

**Comment by LewisCharlesBaker** [comment]
Just bumping this. The issue should be reopened, hundreds of redundant branches in the UI is pretty messy IMO

**Comment by farodyne** [comment]
Past the middle of 2023 and this is still an issue? Will it ever be resolved?

**Comment by rjsdotorg** [comment]
+1 for a button to purge local gone. 
I did (Win7)
`git branch -vv | grep ': gone]'| grep -v "\*" | gawk '{ print $1; }' | xargs -r git branch -d `
and got for example:
```
C:\Users\Ray\Documents\GitHub\quality>git branch -vv | grep ': gone]'| grep -v "\*" | gawk '{ print $1; }' | xargs -r gi
t branch -d
error: The branch 'feature/update' is not fully merged.
If you are sure you want to delete it, run 'git branch -D feature/update'.
error: The branch 'feature/environment_file' is not fully merged.
If you are sure you want to delete it, run 'git branch -D feature/environment_file'.
Deleted branch handle-inputs (was 120fd4f).
```
As desired.

**Comment by emadabdulrahim** [comment]
Still an issue lol

**Comment by nirmalrevarcenturylink** [comment]
Still seeing issue, can you guys able to fix this?

**Comment by kunaall21** [comment]
Still an issue...

**Comment by pawelbabiuch** [comment]
I'm using this command to remove saved local branches that does not exists in repo:

```
git for-each-ref --format="%(refname:short)" refs/heads/* | while read branch; do
    git rev-parse --quiet --verify "refs/remotes/origin/$branch" || git branch -D "$branch"
done
```

**Comment by stiveuk** [comment]
still an issue.

**Comment by njho** [comment]
Still an issue

**Comment by DewRitozz** [comment]
Hello from 2024, still an issue lol

**Comment by nirmalrevarcenturylink** [comment]
Still see same issue, I am using Mac version of Github Desktop.

**Comment by hofmannjan1** [comment]
+1

**Comment by TRaSH-** [comment]
it's closed to they probably don't even see it as a issue, or just don't want to or know how ti fix this issue

**Comment by kale-schuetzeberg** [comment]
This worked for me...

1. **In local directory:**
    ```
    git remote prune origin
    git fetch --prune
    ```

2. **Log out and in of GitHub:** 
   - I didn't need to clear browsing data, but that may help too.

**Comment by nishbarot** [comment]
This is so annoying. It's been 6 years, and there is still no permanent solution for such a simple issue!

**Comment by jmaxwell1-samsung** [comment]
It's utterly ridiculous that this issue still isn't resolved.

**Comment by Distortedlogic** [comment]
Are you tired of the big tech incompetent buffoonery? use better products that are both open source, free, and made by competent devs that actually give a damn.

@jmaxwell1-samsung check this out, its far better than github desktop
https://github.com/Murmele/Gittyup

I still get emails to this old GH acct of mine, but I moved my company completely off github. I think its absolutely clownish that this issue is still open given the resources GH/Microsoft have... absolutely clown world garbage. Idk if anyone else has noticed but more n more of these big tech companies are just getting worse n worse ... if they spent half as much time making the product actually good n work properly instead of tryina get money from me while shoving diversity down my throat in an underhanded way, they would ironically make more money. 

There is an open source free version of almost every big tech software... and the devs that maintain them deserve ur money as a donation for their effort FAR FAR FAR FAR more than these rotten big tech companies. They just are no longer the companies they used to be... just look at the recent stupid move by redis, I moved my company off redis also when that happened.

**Comment by mario-neb** [comment]
this issues discussion is becoming like those nostalgic songs on youtube 🥲 2024 still here?

**Comment by s1radl3y** [comment]
Heyyo, it's 2024 yet here we are lol. Where's the fix lads?

**Comment by complete-dope** [comment]
Still no fix :(

**Comment by sfatew** [comment]
2024, still no fix

**Comment by Flumsykun** [comment]
how is this still an issue!?!?

**Comment by ipsi-apant** [comment]
This is still an issue. It even appearing on a repository on the github web (github.com). My understanding the desktop app is referring 'ghost' branches from github.com. So this issue is from the source. From my point of view, fixing it anything in desktop app may not resolve it.

**Comment by Totolariflette** [comment]
Crazy it's still not fixed. My 20 non existant branches are anoying.

**Comment by Bego96** [comment]
This has to be annoying bug in VS Code. Seriously fix this shit.

**Comment by neerajydv23** [comment]
It's been 6 years guys. Still not fixed!!

**Comment by Saganeiro** [comment]
It's been 6 years guys. Still not fixed!!

**Comment by samparks** [comment]
(Manual) Solution tested on MacOS Github Desktop:

1. Checkout the branch that you've deleted on remote using Github Desktop
2. Click "Branch" at the top of the screen
3. Click "Delete"

This will remove the branch locally and it will no longer appear in Github Desktop.

![image](https://github.com/user-attachments/assets/bf5531ba-40df-41b8-b534-9aacf00c2728)

**Comment by nishbarot** [comment]
> (Manual) Solution tested on MacOS Github Desktop:
> 
> 
> 
> 1. Checkout the branch that you've deleted on remote using Github Desktop
> 
> 2. Click "Branch" at the top of the screen
> 
> 3. Click "Delete"
> 
> 
> 
> This will remove the branch locally and it will no longer appear in Github Desktop.
> 
> 
> 
> ![image](https://github.com/user-attachments/assets/bf5531ba-40df-41b8-b534-9aacf00c2728)
> 
> 

The solution works, but from a developer's perspective, manually performing this task isn't ideal, especially when working with multiple branches and having team members regularly delete branches. Come on GitHub get your act together!

**Comment by mustafaabobakr** [comment]
# Why is this closed?

**Comment by emmanuelmathieu** [comment]
Yes why, there is still the issue

**Comment by jusodi** [comment]
**Please fix this.**
I'm not sure what is more annoying. The issue itself, or the fact that it was highlighted **6 YEARS** ago and nothing has been done to resolve it.

EDIT: The least painful workaround for this that I have found is to delete the local copy, run Github Desktop and re-clone.

**Comment by andreastart2impact** [comment]
A solution for me was to empty the _origin_ folder starting from the repo folder, inside .git/refs/remotes

**Comment by naomi-yu** [comment]
Still an issue. Please fix this.

**Comment by nickpunt** [comment]
This is still an issue github. This creates an error-prone situation for the user to manually manage GitHub Desktop branches which may result in remote branches being accidentally deleted. Any UX confusion that results in potential data loss should be a higher priority.

**Comment by Berend-Kalberg** [comment]
This should be re-opened. This is still an ongoing issue.

**Comment by emilitogaza** [comment]
Would be awesome if this is fixed, makes using the desktop app very painful

**Comment by piyush-nikam** [comment]
Is there an ETA for the fix?

**Comment by sachalchandio** [comment]
why is this still not fixed ? and why is this closed it is 2025 ?

**Comment by CicerBro** [comment]
7 years in and this is still not fixed lol

**Comment by alexiaw111** [comment]
Please fix!!!

**Comment by YuzuGames** [comment]
Still not lol

**Comment by aryabawanta** [comment]
> [@fluffy-cakes](https://github.com/fluffy-cakes) thanks, but doesn't seem to work on mac:
> 
> ```
> git branch -vv | grep ': gone]'|  grep -v "\*" | awk '{ print $1; }' | xargs -r git branch -d
> xargs: illegal option -- r
> usage: xargs [-0opt] [-E eofstr] [-I replstr [-R replacements]] [-J replstr]
>              [-L number] [-n number [-x]] [-P maxprocs] [-s size]
>              [utility [argument ...]]
> ```

Been using this solution for 4 years now, seems github doesn't want to fix this

**Comment by aryabawanta** [comment]
> [@I-Valchev](https://github.com/I-Valchev) macos terminal doesn't need the -r flag: `git branch -vv | grep ': gone]'| grep -v "\*"| awk '{print $1}'| xargs git branch -D ` should work (note the -D flag: this will force the branches to be deleted). This works for me on zsh.

I've been using this solution for like 3 years, seems github doesn't want to fix the issue

**Comment by nishbarot** [comment]
Opened in 2018, ongoing issue in 2025. Simply incompetence from GitHub. Regardless, I created another topic as this one is closed. https://github.com/desktop/desktop/issues/20291. Put your thoughts there, hopefully, we can get this resolved in 2025.

**Comment by niik** [comment]
This issue seems to have transformed into mostly consisting of "mee too"-like comments which doesn't benefit anyone so I'll attempt to explain what GitHub Desktop does and doesn't do when it comes to deleting branches and then lock this thread. There seems to be a fair bit of confusion around who does what and what the difference is between a local and a remote branch so I'll go into some technical detail first but the 

**TL;DR is that GitHub Desktop will remove local branches that have been merged into the default branch as long as they've not been checked out in the last two weeks** 

If you, like me, have used other (centralized) version control systems in the past it's a reasonable expectation that deleting a branch "on the server" would delete it "on the clients" as well. Git however is at its core a decentralized version control system where each user is responsible for pulling changes from other users' repositories.

When cloning or fetching a repository (let's assume it's setup as the `origin` remote) a copy of that repository's branches are saved in the `refs/remotes/origin/` namespace. So if the remote repository has a `feature-a` branch it now exists in your local repository as `refs/remotes/origin/feature-a`. When you check out that branch either by running `git checkout feature-a` or by selecting it in GitHub Desktop a local branch gets created on your machine as `refs/heads/feature-a` and an association between your local repository and the remote is stored in your `.git/config` file. This is what's referred to by Git as a tracking branch.

When GitHub Desktop fetches changes from a remote repository we do so by running `git fetch --prune` which will remove the remote branches that no longer exist. So if the `feature-a` branch gets deleted on the remote then so does `refs/remotes/origin/feature-a`. What's left though is your local branch (`refs/heads/feature-a`) which is now tracking a non-existing upstream branch.

Assuming that the changes in your local `feature-a` branch have been merged into your default branch GitHub Desktop will consider the branch eligible for pruning (i.e. eligible for automated removal). To ensure that we don't delete a branch that users expect to be able to return to however we will not delete it if it's been checked out locally in the last 14 days. We will re-evaluate eligibility for deletion every 24 hours. The exact technical details can be found in https://github.com/desktop/desktop/blob/development/app/src/lib/stores/helpers/branch-pruner.ts.

So to summarize, GitHub Desktop does automatically delete _merged_ local branches that's not tracking a branch that still exists on the remote and that hasn't been checked out in the last two weeks and that's not currently checked out itself. Being able to [manually prune branches](https://github.com/desktop/desktop/issues/13310) is a perfectly reasonable feature request as is (though to a lesser extent) being able to configure Desktop to be much more aggressive in its pruning schedule.

I recommend the chapter on branching in the (free) [Pro Git book](https://git-scm.com/book/en/v2/Git-Branching-Branches-in-a-Nutshell) for those who wishes to learn more.
